src/crab/fwd_analyzer.cpp

```diff
@@ -116,24 +116,27 @@ class interleaved_fwd_fixpoint_iterator_t final {
 
     void operator()(const std::shared_ptr<wto_cycle_t>& cycle);
 
-    friend std::pair<invariant_table_t, invariant_table_t> run_forward_analyzer(cfg_t& cfg, ebpf_domain_t entry_inv);
+    friend std::pair<invariant_table_t, invariant_table_t> run_forward_analyzer(const cfg_t& cfg, ebpf_domain_t entry_inv);
 };
 
-std::pair<invariant_table_t, invariant_table_t> run_forward_analyzer(cfg_t& cfg, ebpf_domain_t entry_inv) {
+std::pair<invariant_table_t, invariant_table_t> run_forward_analyzer(const cfg_t& cfg, ebpf_domain_t entry_inv) {
     // Go over the CFG in weak topological order (accounting for loops).
     interleaved_fwd_fixpoint_iterator_t analyzer(cfg);
     if (thread_local_options.check_termination) {
-        std::vector<label_t> cycle_heads;
-        for (auto& component : analyzer._wto) {
-            if (const auto pc = std::get_if<std::shared_ptr<wto_cycle_t>>(&component)) {
-                cycle_heads.push_back((*pc)->head());
+        // For every potential loop header, initialize the loop counter to 0.
+        for (const auto& label : cfg.potential_loop_headers()) {
+            // The initial path adds labels in the form of (from, to) where
+            // where from is the PC and to is -1. When the CFG is expanded to be non-deterministic
+            // branches with actual to labels are added, but these will not have loop counters as
+            // loop counters are only added to the deterministic cfg.
+            // Ignore labels added by the non-deterministic expansion.
+            if (label.to != -1) {
+                continue;
             }
-        }
-        for (const label_t& label : cycle_heads) {
             entry_inv.initialize_loop_counter(label);
```

Comment:
This should also be an instruction. Not sure if it's better done now or in a future PR.

Reply:
Added a nested loop test case that seems to do the right thing. Let me know what scenarios you are thinking off.